feat: make tokenizer filters configurable

aalexandrov opened the pull request:
# Ticket(s) Closed

- Closes #1565

## What

Implements the syntax extensions proposed in #1565.

## Why

The feature is requested in #1565.

## How

This is done by introducing a new `SearchTokenizerFilters` struct and making this struct a field of all `SearchTokenizer` variants. So far I've only added support for the `default` tokenizer type. If we agree on the general approach I'll add the required code for the remaining variants.

For backwards-compatibility reasons, in the absence of specific layer configuration parameters `lowercase` and `remove_long` the current defaults (`true` and `255`) are assumed in `SearchTokenizerFilters::{lowercase, remove_long_filter}`.

Note that the current WIP relies on some additions to the `tantivy` fork which can be seen here: aalexandrov/tantivy@9be41fcd9aeda654f3d0d902897aac38cd732efb. We can implement this without the `Option<F>` implementation of `TokenFilter`, but then we have to resort to `TextAnalyzerBuilder::{filter_dynamic, dynamic}` which are deemed to be less efficient due to the additional boxing. In general it would be good to run some benchmarks to compare the regressions that we would see from the two approaches, as what am I claiming to be more efficient at the moment is just an educated guess (I guess the goto doc for that is `./cargo-paradedb/README.md`). If we decide to stick with the current approach we'll have to merge the changes linked above into paradedb/tantivy before this PR.

## Tests

I'm still adding them. Please let me know the right place to add an integration test.

aalexandrov committed:
feat: make tokenizer filters configurable WIP

Implements the syntax extensions proposed in #1565.

This is done by introducing a new `SearchTokenizerFilters` struct and
making this struct a field of all `SearchTokenizer` variants.

For backwards-compatibility reasons, in the absence of specific layer
configuration parameters `lowercase` and `remove_long` the current
defaults (`true` and `255`) are assumed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/paradedb/paradedb?pullRequest=1583) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/paradedb/paradedb?pullRequest=1583) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: rebasedming<br/>:x: aalexandrov<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/paradedb/paradedb?pullRequest=1583) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/paradedb/paradedb?pullRequest=1583) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/paradedb/paradedb?pullRequest=1583) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/paradedb/paradedb?pullRequest=1583) it.</sub>

eeeebbbbrrrr commented:
I know this is still WIP, but may I ask why it relies on a custom tantivy fork?

aalexandrov commented:
@eeeebbbbrrrr I've written a section in the "How" as I'm trying out to figure out how to proceed:

> Note that the current WIP relies on some additions to the `tantivy` fork which can be seen here: [aalexandrov/tantivy@9be41fc](https://github.com/aalexandrov/tantivy/commit/9be41fcd9aeda654f3d0d902897aac38cd732efb). We can implement this without the `Option<F>` implementation of `TokenFilter`, but then we have to resort to `TextAnalyzerBuilder::{filter_dynamic, dynamic}` which are deemed to be less efficient due to the additional boxing. In general it would be good to run some benchmarks to compare the regressions that we would see from the two approaches, as what am I claiming to be more efficient at the moment is just an educated guess. If we decide to stick with the current approach we'll have to merge the changes linked above into paradedb/tantivy before this PR.

aalexandrov commented:
@rebasedming / @neilyio: to reiterate, this is currently blocked in WIP you help me resolve the following questions:

1. Is everybody happy with the `SearchTokenizerFilters` approach?
2. What to do about the layer configuration syntax in tantivy? Option (A)—merge the commit linked above in the paradedb fork; option (B)—use the existing (most likely less efficient) API based on boxed layers; option (C)—implement both approaches and benchmark them with `cargo-paradedb` to compare the relative performance penalty and make a more informed decision.
3. Where to integrate the SQL tests from `pg_search/issue_1565.sql` as an end-to-end test? Are there other tests that you definitely want to see?

rebasedming commented:
> @rebasedming / @neilyio: to reiterate, this is currently blocked in WIP until you help me resolve the following questions:
> 
> 1. Is everybody happy with the `SearchTokenizerFilters` approach as a mechanism for configuring and instantiating tokenizer filters?
> 2. What to do about the layer configuration syntax restrictions in tantivy? Option (A)—merge the commit with the `Option<F>` filter implementation linked above in the paradedb fork; option (B)—use the existing (but most likely less efficient) API based on boxed filter layers; option (C)—implement both approaches and benchmark them with `cargo-paradedb` to compare the relative performance penalty and make a more informed decision.
> 3. Where should I integrate the SQL tests from `pg_search/issue_1565.sql` as an end-to-end test (if there is sqllogictest support that might be the best place)? Are there other kinds of tests that you definitely want to see?

Apologies for missing this!

I've run your branch and the approach makes sense. I see why the fork is needed -- having `filter` accept an `Option<F>` seems like the sensible thing to do here. If Tantivy warns that `dynamic` introduces performance penalties, it's safe to listen to them and avoid it.

By the way, I noticed that you only made the lowercaser optional. For consistency let's make the remove_long filter optional as well.

For testing, you'll see that we have unit tests for some of our tokenizers as well as integration tests in the `tests/` folder. For this PR both tests would be useful, since we want to verify that the underlying filters work as expected + we've exposed the APIs correctly.

aalexandrov commented:
> I see why the fork is needed -- having filter accept an Option<F> seems like the sensible thing to do here. If Tantivy warns that dynamic introduces performance penalties, it's safe to listen to them and avoid it.

I opened a PR for that in paradedb/tantivy#4. We need to merge it first so I can fix the `Cargo.toml` changes here before we merge this PR.

aalexandrov commented:
> By the way, I noticed that you only made the lowercaser optional. For consistency let's make the remove_long filter optional as well.

The code was originally

```rust
fn remove_long_filter(&self) -> RemoveLongFilter {
    let limit = self.remove_long.unwrap_or(DEFAULT_REMOVE_TOKEN_LENGTH);
    RemoveLongFilter::limit(limit)
}
```

and after the requested changes now looks like this.

```rust
fn remove_long_filter(&self) -> Option<RemoveLongFilter> {
    let limit = self.remove_long.unwrap_or(DEFAULT_REMOVE_TOKEN_LENGTH);
    Some(RemoveLongFilter::limit(limit))
}
```

The result value is always going to be `Some(filter)` because at the moment there is no way to completely disable this filter—a client configuration either sets the `remove_long` value explicitly as part of the tokenizer definition, or we resort to a filter with the default value.

An implementation that is more consistent with the `LowerCaser` constructor but this is not backwards-compatible would look like this:

```rust
fn remove_long_filter(&self) -> Option<RemoveLongFilter> {
    match self.lowercase {
        Some(limit) => Some(RemoveLongFilter::limit(limit)),
        _ => None,
    }
}
```

aalexandrov commented:
> For testing, you'll see that we have unit tests for some of our tokenizers as well as integration tests in the `tests/` folder. For this PR both tests would be useful, since we want to verify that the underlying filters work as expected + we've exposed the APIs correctly.

* I fixed some tests in `manager::tests` to cover the new `SearchTokenizerFilters` struct.
* I added a `tokenizer_filters` in `pg_search/tests/tokenize.rs`.

aalexandrov commented:
@rebasedming I think this should be ready for review, although merging is currently blocked by both paradedb/paradedb#1618 and paradedb/tantivy#4 (in that order). See my comments above for details.

The changes here are all backwards compatible. Overall I think that once layer configuration is introduced one can rethink the set of layers exposed to the system and the default values for the various layers. For example I'm not sure whether the `lowercase` type is still needed. However, I suggest to discuss whether a more obvious, backwards-incompatible revamp of the mapping between the current `tokenizer` SQL-level API and the tantivy API is needed in a separate issue in order to avoid scope creep.

philippemnoel commented:
> @rebasedming I think this should be ready for review, although merging is currently blocked by both #1618 and [paradedb/tantivy#4](https://github.com/paradedb/tantivy/pull/4) (in that order). See my comments above for details.
> 
> The changes here are all backwards compatible. Overall I think that once layer configuration is introduced one can rethink the set of layers exposed to the system and the default values for the various layers. For example I'm not sure whether the `lowercase` type is still needed. However, I suggest to discuss whether a more obvious, backwards-incompatible revamp of the mapping between the current `tokenizer` SQL-level API and the tantivy API is needed in a separate issue in order to avoid scope creep.

Thank you! We'll get the Tantivy rebase in first and come to this PR after.

philippemnoel commented:
> @rebasedming I think this should be ready for review, although merging is currently blocked by both #1618 and [paradedb/tantivy#4](https://github.com/paradedb/tantivy/pull/4) (in that order). See my comments above for details.
> 
> The changes here are all backwards compatible. Overall I think that once layer configuration is introduced one can rethink the set of layers exposed to the system and the default values for the various layers. For example I'm not sure whether the `lowercase` type is still needed. However, I suggest to discuss whether a more obvious, backwards-incompatible revamp of the mapping between the current `tokenizer` SQL-level API and the tantivy API is needed in a separate issue in order to avoid scope creep.

Your Tantivy PR has been merged and the rebase PR is ready. Once we merge it this will be unblocked. Thank you for your patience!

rebasedming committed:
use paradedb tantivy fork

rebasedming committed:
use paradedb tantivy fork

philippemnoel commented:
> > @rebasedming I think this should be ready for review, although merging is currently blocked by both #1618 and [paradedb/tantivy#4](https://github.com/paradedb/tantivy/pull/4) (in that order). See my comments above for details.
> > The changes here are all backwards compatible. Overall I think that once layer configuration is introduced one can rethink the set of layers exposed to the system and the default values for the various layers. For example I'm not sure whether the `lowercase` type is still needed. However, I suggest to discuss whether a more obvious, backwards-incompatible revamp of the mapping between the current `tokenizer` SQL-level API and the tantivy API is needed in a separate issue in order to avoid scope creep.
> 
> Your Tantivy PR has been merged and the rebase PR is ready. Once we merge it this will be unblocked. Thank you for your patience!

This PR is fully unblocked @aalexandrov.

rebasedming committed:
rebase

rebasedming committed:
add test
